packages/prisma-client/prisma/migrations/20221218211129_tree_text/migration.ts

```diff
@@ -0,0 +1,64 @@
+import { PrismaClient } from "./client";
+
+type PrevInstance = {
+  id: string;
+  component: any;
+  cssRules: Array<any>;
+  children: Array<PrevInstance | string>;
+};
+
+type Text = {
+  type: "text";
+  value: string;
+};
+
+type Instance = {
+  type: "instance";
+  id: string;
+  component: any;
+  cssRules: Array<any>;
+  children: Array<Instance | Text>;
+};
+
+const convertTree = (prevInstance: PrevInstance): Instance => {
+  const instance: Instance = {
+    type: "instance",
+    id: prevInstance.id,
+    component: prevInstance.component,
+    cssRules: prevInstance.cssRules,
+    children: [],
+  };
+
+  for (const child of prevInstance.children) {
+    if (typeof child === "string") {
+      instance.children.push({ type: "text", value: child });
+    } else {
+      instance.children.push(convertTree(child));
+    }
+  }
+  return instance;
+};
+
+export default () => {
+  const client = new PrismaClient({
+    // Uncomment to see the queries in console as the migration runs
+    // log: ["query", "info", "warn", "error"],
+  });
+  return client.$transaction(
+    async (prisma) => {
+      const previousTrees = await prisma.tree.findMany();
```

Comment:
On prod we have 8855 Trees. IMHO this will definitely will timeout. Have you tried copy prod db and try locally?

Reply:
btw what do you think we could do to make it work with very large tables? Some kind of batching?

Reply:
sql would be fast if stored with native types instead of json

Reply:
Btw where all these pages comes from? From test cases?

Reply:
sql is not the only problem, you are fetching the entire table there and in the real world it will be many gigabytes some day that you are trying to put into memory

Reply:
> Btw where all these pages comes from? From test cases?

in prod - yeah, from people clicking on "try" button and we clone for them the landing project

Reply:
> sql is not the only problem, you are fetching the entire table there and in the real world it will be many gigabytes some day that you are trying to put into memory

Yes, I know. Solution is not the best. Didn't expect we have 8k trees already.
Does prisma have any tools for this?

Reply:
Increase timeout maybe, what time now does it take?

Reply:
One of options to update as a batch is to use raw query and `unnest`, most raws allows to pass array as params
```
  UPDATE blabla AS t SET
    name = v.new_name
    blabbla = v.new_blabla
  FROM unnest(?::text[], ?::text[]) AS v(id, new_name, new_blabla)
  WHERE t.id = v.id
```

Reply:
We probably need a pagination helper that would allow to fetch X amount of items from the large table, then another batch and another batch

Reply:
Yeah, lots of ways to write migrations more efficiently if we use raw SQL in some ways. But I couldn't convince @kof that we should just use raw SQL when makes sense 😅

Reply:
In case of db.raw don't understand arrays as params:
you can inject array into string.
i.e. instead of this `?::uuid[]` inject `array['id1', 'id2', '...']` etc but you need to quote each value, so better to use escape lib like 
https://www.npmjs.com/package/sqlstring

Reply:
Also, sending 8000 `.update()` queries shouldn't take more than 5 minutes in prod with current setup, probably 1-2 minutes. So we don't have to overthink this particular migration too much. Should work as is.

Reply:
> Yeah, lots of ways to write migrations more efficiently if we use raw SQL in some ways. But I couldn't convince @kof that we should just use raw SQL when makes sense 😅

I just know that complex data migrations using just SQL would limit everyone. Only @istarkov and a few other people on the planet know how to write those for queries that require json parse and various other js manipulations.

without this migration tool we have now, even this migration here wouldn't be already done (I bet it took less than a few hours to write)

For speed of development this is huge and we will find ways to make them run well. That raw SQL stuff for complex migrations is not an option right now. Only if we make the entire data completely normalized and the query becomes human readable.

Reply:
$ yarn migrations --dev migrate
yarn run v1.22.19
$ /Users/bogdancadkin/host/webstudio-designer/node_modules/.bin/migrations --dev migrate
You're about to apply the following migration(s):

  - 20221218211129_tree_text

? Continue? yes
Starting 20221218211129_tree_text
Tree 621536ed2e8214a621cbdf5e cannot be converted
Tree 628513c2bd1a04feefcd32c4 cannot be converted
Tree 628513d2bd1a04feefcd32c6 cannot be converted
Tree 628513d6bd1a04feefcd32c8 cannot be converted
20221218211129_tree_text done in 69.886s
✨  Done in 72.93s.

Reply:
> Tree 621536ed2e8214a621cbdf5e cannot be converted

These messages are from errored rows

Reply:
> I just know that complex data migrations using just SQL would limit everyone. Only @istarkov and a few other people on the planet know how to write those for queries that require json parse and various other js manipulations.

Yeah. Doing it completely in an SQL file would be tough. But now that we have .ts migrations, I think using some raw SQL in the .ts files is fine. E.g. to combine many `update` queries into one as @istarkov have suggested above. And the SQL he posted is very readable to me, so that makes at least two :)

Reply:
ah, yeah, that can be done from time to time if need, I thought the discussion is about generally writing migrations with ts